Unstuck car implementation

Hornwitser opened the pull request:
Change to using an assembly machine as the name for the find_non_colliding_possition call.  This will make it use a larger bounding box for checking collisions and thereby place the car some distance away from the wall.

Implement /unstuck command as detailed in #51.  No GUI since I'm not sure how to do that properly.  The command might be a bit abusable as it generally places you some distance away from your current position, so if you collide into an obstacle and use it you get put left or right from your current position and could probably start driving immediately.  Same goes for escaping a biter swarm with it.

Hornwitser committed:
Use assembler for finding new car position

Change the entity name used to find a non-colliding position to
assembling-machine-1.  This moves the car away from close obstacles it
could get stuck in due to the non-colliding position algorithm not
taking account to the rotation of the car.

Hornwitser committed:
Add /unstuck command

Add unstuck command for the Race game which moves the player's car away
from nearby obstacles.

Hornwitser committed:
Fix mini-games commands not being disabled

Fix the commands part of mini-games not being disabled when the game is
over.

tovernaar123 commented on modules/mini-games/Race.lua:
```diff
@@ -413,11 +414,24 @@ local kill_biters = Token.register(function(name)
 respawn_car = Token.register(function(name)
     local player = dead_cars[name].player
-    local position = surface[1].find_non_colliding_position('car', dead_cars[name].position, 5, 0.5)
+    local position = surface[1].find_non_colliding_position('assembling-machine-1', dead_cars[name].position, 5, 0.5)
```
mabye just use smt like rocket silo so you dont need the command at all

Hornwitser replied:
Rocket silo is 9x9.  That would make it fail in narrow sections and is completely unacceptable.  And even so, it's still possible to get pretty stuck driving parallel to a straight section of wall so you still need the command anyhow.

Cooldude2606 replied:
The assembling seams like the right choice as its about the same size as a car but without the rotational difference.

tovernaar123 commented on modules/mini-games/Race.lua:
```diff
@@ -413,11 +414,24 @@ local kill_biters = Token.register(function(name)
+        return Commands.error
+    end
+
+    local new_position = surface[1].find_non_colliding_position('assembling-machine-1', car.position, 5, 0.5)
+    car.teleport(new_position, surface[1])
```
this still wont work as find_non_colliding_position does not care about rotation

Hornwitser replied:
An assembling machine is 3x3.  You can drive perfectly fine when placed in the centre of a position where an assembling machine fits.

tovernaar123 replied:
if this would be the case the command would not be needed but as pointed out above even with a 9 x 9 shape "And even so, it's still possible to get pretty stuck driving parallel to a straight section of wall so you still need the command anyhow."

tovernaar123 replied:
as the command is doing the same as the car place logic

Hornwitser replied:
Yes.  Drive parallel to a wall and turn and you get stuck.  You need the command to get unstuck or end up spending minutes fiddling around to get free.

tovernaar123 replied:
again its precisly the same as respawing the car so idk what it will do ? all it will do is try to find a non_colliding_position which crashing a car does the same (which is the problem with really getting stuck)

Hornwitser replied:
Also what part of being placed in the centre of a 3x3 non-colliding area guarantees your car is not stuck is it that you don't understand?

tovernaar123 replied:
no what i dont get is that this is the same is respwning the cars which is the problem right?

Hornwitser replied:
The cars no longer respawn inside walls.  That's what change from car to assembling-machine-1 does.

tovernaar123 replied:
yes so why is this command needed?

Hornwitser replied:
As I told you three times now if you drive close to a straight section of a wall an turn wrong you get stuck against the wall.

tovernaar123 replied:
yes but youre not really stuck sutck youre not colliding so wont this just return itslef?

tovernaar123 replied:
i mean its called find_non_colliding_position and youre not collind so i think you wont really tp any where

Hornwitser replied:
Can you fit a 3x3 tile sized assembler in the position where the car is when close up to the wall?  No!  The wall is in the way, and the car is in the way.

Why don't you test what you think is a problem and confirm that it is one _before_ you comment with speculation on your part that is simply wrong.

tovernaar123 replied:
o wait ofc the car is still their srry then you spam this command to get out of the walls (mabye ) also i cant test rn i am on mobile sorry

Hornwitser replied:
Even if you pileup three cars around on a narrow wall like this
![image](https://user-images.githubusercontent.com/14362102/91296391-d3a5e300-e79c-11ea-8648-c48bd7330db5.png)

You're still not going to teleport through the wall with /unstuck
![image](https://user-images.githubusercontent.com/14362102/91296455-eae4d080-e79c-11ea-8902-f7af0417259c.png)

Sure, if you got like 5, 6, 7 cars around preventing any positions nearby then yeah, you'd probably be able to teleport through the wall.  But that's enough coordination to warrant the players involved being kicked for breaking the game.

tovernaar123 replied:
ok 👌 good srry for syaing this then i really thought it was wrong

tovernaar123 commented on modules/mini-games/Race.lua:
```diff
@@ -413,11 +414,24 @@ local kill_biters = Token.register(function(name)
+:register(function(player)
+    local car = cars[player.name]
+    if not car or not car.valid then
+        player.print("Not in a car!")
+        return Commands.error
```
this can be return Commands.error("Not in a car!")

Hornwitser replied:
Ok.

tovernaar123 commented on expcore/Mini_games.lua:
```diff
@@ -820,7 +820,7 @@ function Mini_games.stop_game()
     -- Disable all commands for this mini game
     for _, command_name  in ipairs(mini_game.commands) do
-        Commands.enable(command_name)
+        Commands.disable(command_name)
```
nice bug fix 👍

Hornwitser committed:
Pass error to Command.error

Pass the missing car error message to Command.error and let the command
system print it to the user.

tovernaar123 reviewed (approved):
might make it a gui latter but this is very nice for now

Cooldude2606 reviewed (approved):
Everything looks good, althouh I would suggest some sort of timeout rather than a gui for doing the unstuck. Meaning if you stop moving for X seconds and you are close to a wall then it will unstuck you.
